rkt/fetch.go

```diff
@@ -96,3 +51,70 @@ func runFetch(args []string) (exit int) {
 
 	return
 }
+
+// fetchImage will take an image as either a URL or a name string and import it
+// into the store if found.
+func fetchImage(img string, ds *cas.Store) (string, error) {
+	var u *url.URL
+	var err error
+	if app, ok := findAppString(img); ok {
+		fmt.Printf("rkt: starting to discover app img %s\n", img)
+		u, err = discover(app)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		u, err = url.Parse(img)
+	}
+	if err != nil {
+		return "", fmt.Errorf("not a valid URL (%s)", img)
+	}
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return "", fmt.Errorf("rkt only supports http or https URLs (%s)", img)
+	}
+	return fetchURL(img, ds)
+}
+
+func fetchURL(img string, ds *cas.Store) (string, error) {
+	fmt.Printf("rkt: starting to fetch img from %s\n", img)
+	rem := cas.NewRemote(img, []string{})
+	err := ds.ReadIndex(rem)
+	if err != nil && rem.Blob == "" {
+		rem, err = rem.Download(*ds)
+		if err != nil {
+			return "", fmt.Errorf("fetch: %v\n", err)
+		}
+	}
+	return rem.Blob, nil
+}
+
+func discover(app *discovery.App) (*url.URL, error) {
+	ep, err := discovery.DiscoverEndpoints(*app, true)
+	if err != nil {
+		return nil, fmt.Errorf("discovery: %v", err)
+	}
+	// TODO(philips): use all available mirrors
+	if globalFlags.Debug {
+		fmt.Printf("discovery: trying %v\n", ep.ACI[0])
+	}
+	imgurl := ep.ACI[0]
+	u, err := url.Parse(imgurl)
+	if err != nil {
+		return nil, fmt.Errorf("discovery: fetched img URL (%s) is invalid", imgurl)
+	}
+	return u, nil
+}
+
+// findAppString returns the appString and true if the img is an app name and
```

Comment:
@xiang90 I think "findApp" is a bit confusing because it suggests that it might actually perform some kind of discovery, whereas it's actually just trying to create a new struct based on the given name